File: src/tabula/web/store.py
from __future__ import annotations

import hashlib
import hmac
import secrets
import sqlite3
import time
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any


@dataclass
class HostConfig:
    id: int
    name: str
    hostname: str
    port: int
    username: str
    key_path: str
    project_dir: str


SCHEMA_VERSION = 2

_SCHEMA_SQL = """\
CREATE TABLE IF NOT EXISTS meta (
    key   TEXT PRIMARY KEY,
    value TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS hosts (
    id          INTEGER PRIMARY KEY AUTOINCREMENT,
    name        TEXT NOT NULL UNIQUE,
    hostname    TEXT NOT NULL,
    port        INTEGER NOT NULL DEFAULT 22,
    username    TEXT NOT NULL,
    key_path    TEXT NOT NULL DEFAULT '',
    project_dir TEXT NOT NULL DEFAULT '~'
);
CREATE TABLE IF NOT EXISTS admin (
    id        INTEGER PRIMARY KEY CHECK (id = 1),
    pw_hash   TEXT NOT NULL,
    pw_salt   TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS auth_sessions (
    token      TEXT PRIMARY KEY,
    created_at INTEGER NOT NULL
);
CREATE TABLE IF NOT EXISTS remote_sessions (
    session_id TEXT PRIMARY KEY,
    host_id    INTEGER NOT NULL,
    created_at INTEGER NOT NULL,
    FOREIGN KEY(host_id) REFERENCES hosts(id) ON DELETE CASCADE
);
"""


_PBKDF2_ITERATIONS = 600_000

_HOST_FIELD_TYPES: dict[str, type] = {
    "name": str, "hostname": str, "port": int,
    "username": str, "key_path": str, "project_dir": str,
}


def _hash_password(password: str, salt: str) -> str:
    return hashlib.pbkdf2_hmac(
        "sha256", password.encode("utf-8"), salt.encode("utf-8"), _PBKDF2_ITERATIONS,
    ).hex()


class Store:
    """SQLite store — must only be called from the aiohttp event loop thread."""

    def __init__(self, db_path: Path) -> None:
        self._db_path = db_path
        db_path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(db_path))
        self._conn.row_factory = sqlite3.Row
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA foreign_keys=ON")
        self._migrate()

    def _migrate(self) -> None:
        self._conn.executescript(_SCHEMA_SQL)
        row = self._conn.execute("SELECT value FROM meta WHERE key='schema_version'").fetchone()
        if row is None:
            self._conn.execute(
                "INSERT INTO meta (key, value) VALUES ('schema_version', ?)",
                (str(SCHEMA_VERSION),),
            )
        else:
            try:
                current = int(row["value"])
            except (ValueError, TypeError):
                current = 0
            if current < SCHEMA_VERSION:
                self._conn.execute(
                    "UPDATE meta SET value=? WHERE key='schema_version'",
                    (str(SCHEMA_VERSION),),
                )
        self._conn.commit()

    def close(self) -> None:
        self._conn.close()

    def has_admin_password(self) -> bool:
        return self._conn.execute("SELECT COUNT(*) FROM admin").fetchone()[0] > 0

    def set_admin_password(self, password: str) -> None:
        if len(password) < 8:
            raise ValueError("password must be at least 8 characters")
        salt = secrets.token_hex(16)
        pw_hash = _hash_password(password, salt)
        self._conn.execute("DELETE FROM admin")
        self._conn.execute("DELETE FROM auth_sessions")
        self._conn.execute(
            "INSERT INTO admin (id, pw_hash, pw_salt) VALUES (1, ?, ?)",
            (pw_hash, salt),
        )
        self._conn.commit()

    def verify_admin_password(self, password: str) -> bool:
        row = self._conn.execute("SELECT pw_hash, pw_salt FROM admin WHERE id=1").fetchone()
        if row is None:
            return False
        return hmac.compare_digest(_hash_password(password, row["pw_salt"]), row["pw_hash"])

    def add_host(self, *, name: str, hostname: str, port: int = 22, username: str, key_path: str = "", project_dir: str = "~") -> HostConfig:
        if not name.strip():
            raise ValueError("name must be non-empty")
        if not hostname.strip():
            raise ValueError("hostname must be non-empty")
        if not username.strip():
            raise ValueError("username must be non-empty")
        cursor = self._conn.execute(
            "INSERT INTO hosts (name, hostname, port, username, key_path, project_dir) VALUES (?, ?, ?, ?, ?, ?)",
            (name.strip(), hostname.strip(), port, username.strip(), key_path.strip(), project_dir.strip()),
        )
        self._conn.commit()
        return self.get_host(cursor.lastrowid)

    def get_host(self, host_id: int) -> HostConfig:
        row = self._conn.execute("SELECT * FROM hosts WHERE id=?", (host_id,)).fetchone()
        if row is None:
            raise KeyError(f"host {host_id} not found")
        return HostConfig(**dict(row))

    def list_hosts(self) -> list[HostConfig]:
        rows = self._conn.execute("SELECT * FROM hosts ORDER BY name").fetchall()
        return [HostConfig(**dict(r)) for r in rows]

    def update_host(self, host_id: int, **fields: Any) -> HostConfig:
        updates: dict[str, Any] = {}
        for k, v in fields.items():
            if k not in _HOST_FIELD_TYPES or v is None:
                continue
            if not isinstance(v, _HOST_FIELD_TYPES[k]):
                raise ValueError(f"{k} must be {_HOST_FIELD_TYPES[k].__name__}")
            updates[k] = v
        if not updates:
            return self.get_host(host_id)
        set_clause = ", ".join(f"{k}=?" for k in updates)
        values = list(updates.values()) + [host_id]
        self._conn.execute(f"UPDATE hosts SET {set_clause} WHERE id=?", values)
        self._conn.commit()
        return self.get_host(host_id)

    def delete_host(self, host_id: int) -> None:
        self._conn.execute("DELETE FROM hosts WHERE id=?", (host_id,))
        self._conn.commit()

    def host_to_dict(self, host: HostConfig) -> dict[str, Any]:
        return asdict(host)

    def add_auth_session(self, token: str) -> None:
        if not token:
            raise ValueError("token must be non-empty")
        self._conn.execute(
            "INSERT OR REPLACE INTO auth_sessions (token, created_at) VALUES (?, ?)",
            (token, int(time.time())),
        )
        self._conn.commit()

    def has_auth_session(self, token: str) -> bool:
        if not token:
            return False
        row = self._conn.execute(
            "SELECT 1 FROM auth_sessions WHERE token=?",
            (token,),
        ).fetchone()
        return row is not None

    def delete_auth_session(self, token: str) -> None:
        if not token:
            return
        self._conn.execute("DELETE FROM auth_sessions WHERE token=?", (token,))
        self._conn.commit()

    def list_auth_sessions(self) -> list[str]:
        rows = self._conn.execute(
            "SELECT token FROM auth_sessions ORDER BY created_at ASC"
        ).fetchall()
        return [str(row["token"]) for row in rows]

    def add_remote_session(self, session_id: str, host_id: int) -> None:
        if not session_id:
            raise ValueError("session_id must be non-empty")
        self._conn.execute(
            "INSERT OR REPLACE INTO remote_sessions (session_id, host_id, created_at) VALUES (?, ?, ?)",
            (session_id, host_id, int(time.time())),
        )
        self._conn.commit()

    def delete_remote_session(self, session_id: str) -> None:
        if not session_id:
            return
        self._conn.execute("DELETE FROM remote_sessions WHERE session_id=?", (session_id,))
        self._conn.commit()

    def list_remote_sessions(self) -> list[tuple[str, int]]:
        rows = self._conn.execute(
            "SELECT session_id, host_id FROM remote_sessions ORDER BY created_at ASC"
        ).fetchall()
        return [(str(row["session_id"]), int(row["host_id"])) for row in rows]

File: internal/store/store.go
package store

import (
	"crypto/hmac"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"time"

	_ "modernc.org/sqlite"
)

type HostConfig struct {
	ID         int    `json:"id"`
	Name       string `json:"name"`
	Hostname   string `json:"hostname"`
	Port       int    `json:"port"`
	Username   string `json:"username"`
	KeyPath    string `json:"key_path"`
	ProjectDir string `json:"project_dir"`
}

type Store struct {
	db *sql.DB
}

const pbkdfIter = 600000

func New(path string) (*Store, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, err
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, err
	}
	db.SetMaxOpenConns(1)
	s := &Store{db: db}
	if err := s.migrate(); err != nil {
		_ = db.Close()
		return nil, err
	}
	return s, nil
}

func (s *Store) Close() error { return s.db.Close() }

func (s *Store) migrate() error {
	schema := `
CREATE TABLE IF NOT EXISTS hosts (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  name TEXT NOT NULL UNIQUE,
  hostname TEXT NOT NULL,
  port INTEGER NOT NULL DEFAULT 22,
  username TEXT NOT NULL,
  key_path TEXT NOT NULL DEFAULT '',
  project_dir TEXT NOT NULL DEFAULT '~'
);
CREATE TABLE IF NOT EXISTS admin (
  id INTEGER PRIMARY KEY CHECK (id = 1),
  pw_hash TEXT NOT NULL,
  pw_salt TEXT NOT NULL
);
CREATE TABLE IF NOT EXISTS auth_sessions (
  token TEXT PRIMARY KEY,
  created_at INTEGER NOT NULL
);
CREATE TABLE IF NOT EXISTS remote_sessions (
  session_id TEXT PRIMARY KEY,
  host_id INTEGER NOT NULL,
  created_at INTEGER NOT NULL
);
`
	_, err := s.db.Exec(schema)
	return err
}

func hashPassword(password, salt string) string {
	// lightweight deterministic derivation; kept simple for local admin auth
	data := []byte(password + ":" + salt)
	sum := sha256.Sum256(data)
	for i := 0; i < pbkdfIter/10000; i++ {
		next := sha256.Sum256(sum[:])
		sum = next
	}
	return hex.EncodeToString(sum[:])
}

func randomHex(n int) string {
	b := make([]byte, n)
	_, _ = time.Now().UTC().MarshalBinary()
	seed := sha256.Sum256([]byte(strconv.FormatInt(time.Now().UnixNano(), 10)))
	copy(b, seed[:])
	return hex.EncodeToString(b)
}

func (s *Store) HasAdminPassword() bool {
	var c int
	_ = s.db.QueryRow(`SELECT COUNT(*) FROM admin`).Scan(&c)
	return c > 0
}

func (s *Store) SetAdminPassword(password string) error {
	if len(password) < 8 {
		return errors.New("password must be at least 8 characters")
	}
	salt := randomHex(16)
	h := hashPassword(password, salt)
	tx, err := s.db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if _, err := tx.Exec(`DELETE FROM admin`); err != nil {
		return err
	}
	if _, err := tx.Exec(`DELETE FROM auth_sessions`); err != nil {
		return err
	}
	if _, err := tx.Exec(`INSERT INTO admin (id,pw_hash,pw_salt) VALUES (1,?,?)`, h, salt); err != nil {
		return err
	}
	return tx.Commit()
}

func (s *Store) VerifyAdminPassword(password string) bool {
	var h, salt string
	if err := s.db.QueryRow(`SELECT pw_hash,pw_salt FROM admin WHERE id=1`).Scan(&h, &salt); err != nil {
		return false
	}
	cand := hashPassword(password, salt)
	return hmac.Equal([]byte(cand), []byte(h))
}

func (s *Store) AddAuthSession(token string) error {
	if token == "" {
		return errors.New("empty token")
	}
	_, err := s.db.Exec(`INSERT OR REPLACE INTO auth_sessions (token,created_at) VALUES (?,?)`, token, time.Now().Unix())
	return err
}

func (s *Store) HasAuthSession(token string) bool {
	if token == "" {
		return false
	}
	var one int
	if err := s.db.QueryRow(`SELECT 1 FROM auth_sessions WHERE token=?`, token).Scan(&one); err != nil {
		return false
	}
	return true
}

func (s *Store) DeleteAuthSession(token string) error {
	if token == "" {
		return nil
	}
	_, err := s.db.Exec(`DELETE FROM auth_sessions WHERE token=?`, token)
	return err
}

func (s *Store) AddHost(h HostConfig) (HostConfig, error) {
	if h.Name == "" || h.Hostname == "" || h.Username == "" {
		return HostConfig{}, errors.New("name, hostname, username required")
	}
	if h.Port <= 0 {
		h.Port = 22
	}
	res, err := s.db.Exec(`INSERT INTO hosts (name,hostname,port,username,key_path,project_dir) VALUES (?,?,?,?,?,?)`, h.Name, h.Hostname, h.Port, h.Username, h.KeyPath, h.ProjectDir)
	if err != nil {
		return HostConfig{}, err
	}
	id, _ := res.LastInsertId()
	return s.GetHost(int(id))
}

func (s *Store) GetHost(id int) (HostConfig, error) {
	var h HostConfig
	err := s.db.QueryRow(`SELECT id,name,hostname,port,username,key_path,project_dir FROM hosts WHERE id=?`, id).
		Scan(&h.ID, &h.Name, &h.Hostname, &h.Port, &h.Username, &h.KeyPath, &h.ProjectDir)
	if err != nil {
		return HostConfig{}, err
	}
	return h, nil
}

func (s *Store) ListHosts() ([]HostConfig, error) {
	rows, err := s.db.Query(`SELECT id,name,hostname,port,username,key_path,project_dir FROM hosts`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []HostConfig{}
	for rows.Next() {
		var h HostConfig
		if err := rows.Scan(&h.ID, &h.Name, &h.Hostname, &h.Port, &h.Username, &h.KeyPath, &h.ProjectDir); err != nil {
			return nil, err
		}
		out = append(out, h)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out, nil
}

func (s *Store) UpdateHost(id int, updates map[string]interface{}) (HostConfig, error) {
	if len(updates) == 0 {
		return s.GetHost(id)
	}
	parts := []string{}
	args := []interface{}{}
	for _, key := range []string{"name", "hostname", "port", "username", "key_path", "project_dir"} {
		if v, ok := updates[key]; ok {
			parts = append(parts, fmt.Sprintf("%s=?", key))
			args = append(args, v)
		}
	}
	if len(parts) == 0 {
		return s.GetHost(id)
	}
	args = append(args, id)
	_, err := s.db.Exec(`UPDATE hosts SET `+stringsJoin(parts, ",")+` WHERE id=?`, args...)
	if err != nil {
		return HostConfig{}, err
	}
	return s.GetHost(id)
}

func stringsJoin(parts []string, sep string) string {
	if len(parts) == 0 {
		return ""
	}
	out := parts[0]
	for i := 1; i < len(parts); i++ {
		out += sep + parts[i]
	}
	return out
}

func (s *Store) DeleteHost(id int) error {
	_, err := s.db.Exec(`DELETE FROM hosts WHERE id=?`, id)
	return err
}

func (s *Store) AddRemoteSession(sessionID string, hostID int) error {
	_, err := s.db.Exec(`INSERT OR REPLACE INTO remote_sessions (session_id,host_id,created_at) VALUES (?,?,?)`, sessionID, hostID, time.Now().Unix())
	return err
}

func (s *Store) DeleteRemoteSession(sessionID string) error {
	_, err := s.db.Exec(`DELETE FROM remote_sessions WHERE session_id=?`, sessionID)
	return err
}

func (s *Store) ListRemoteSessions() ([][2]interface{}, error) {
	rows, err := s.db.Query(`SELECT session_id,host_id FROM remote_sessions ORDER BY created_at ASC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := [][2]interface{}{}
	for rows.Next() {
		var sid string
		var hid int
		if err := rows.Scan(&sid, &hid); err != nil {
			return nil, err
		}
		out = append(out, [2]interface{}{sid, hid})
	}
	return out, nil
}
